### PR TITLE
Add per-cpu Page Tables.

### PR DESCRIPTION
**Pull request by glg-rv**
This implementation is rather naive. It creates multiple copies of the same mappings per cpu.

This is a stepping stone to allow per-cpu virtual mapping of user mode areas.

Note: All we'd really need at this point, as all mapping are all shared, is a per-CPU top level page table, and impose that all per-cpu mappings are in a specific area where they only share the top-level pagetable with other mappings -- to avoid refcounting issues on PTEs. This _will_ the case, but might require some delicate changes in the current page tables code, therefore this work is deferred to future Gianluca.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -264,9 +260,9 @@ fn hyp_map_region(
...
 
-// Creates the Sv48 page table based on the accessible regions of memory in the provided memory
-// map.
-fn setup_hyp_paging(mem_map: &mut HwMemMap) {
+/// Creates a Sv48 page table based on the accessible regions of memory in the
+/// provided memory map.
+pub fn setup_hyp_paging(mem_map: &mut HwMemMap) -> FirstStagePageTable<Sv48> {
     let num_pte_pages = pte_page_count(mem_map);
     let pte_base = find_available_region(mem_map, num_pte_pages * PageSize::Size4k as u64)
         .expect("Not enough free memory for hypervisor Sv48 page table");
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/main.rs:268]: Oh. Yes, this is true. Wow the code is fragile. Okay, I seem I need to to this properly and not with a simple thing. In any case, the core of this patch was to suggest to move the creation of page tables at PerCpu init time, so at least it served its purpose.

File: src/smp.rs [hunk 2]
```diff
@@ -56,12 +58,19 @@ impl PerCpu {
             .unwrap();
         PER_CPU_BASE.call_once(|| pcpu_base);
 
-        // Now initialize each PerCpu structure.
         for i in 0..cpu_info.num_cpus() {
             let cpu_id = CpuId::new(i);
+
+            // Initialize each cpu root pagetable.
+            let rootpt = setup_hyp_paging(mem_map);
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/smp.rs:65]: Okay. I was kind of going for a "map page" interface that didn't have to specify the page table (implying current), and get it from SATP. But well, the same can be reached by keeping FirstStagePageTable instead.

[the remark this answers is not in the document]
Comment (glg-rv) [reply src/smp.rs]: Done.

File: src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/main.rs]: Added reserved_hyp_paging. to allocate them at once.

Actually, is the fact that we allocate them in the memory region (so tagging them as hypervisor ptes) useful? Or that's not important in the grand scheme of things?

File: src/smp.rs [hunk 3]
```diff
@@ -25,6 +27,7 @@ extern "C" {
 pub struct PerCpu {
     cpu_id: CpuId,
     vmid_tracker: RefCell<VmIdTracker>,
+    rootpt: FirstStagePageTable<Sv48>,
```
Comment (abrestic-rivos) [inline src/smp.rs:30]: nit: I would just call this `page_table`

File: src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (abrestic-rivos) [reply src/main.rs]: I'm not sure that it is useful for them to be in their own `HwMemMap` region. IMO it's simpler if we treat it like ordinary hypervisor-internal memory that we allocate from `HypPageAlloc`.

**Comment by glg-rv** [comment]
Closing down this, as a new iteration of the patch is ready.

**Pull request closed without merging**